### ISSUE TITLE
The way to get total count when I use paging

### ISSUE THREAD

**Issue by a188037445** [mybatis/mybatis-dynamic-sql#271]
Now I use two queries to do it, first to query total count, second to query real data paged with `limit()` and `offset()`.

But when I use a long "where" clause I have to write them twice in two queries, is there any better ways to improve it?

**Comment by jeffgbutler** [comment]
See this test for an example of reusing a where clause: https://github.com/mybatis/mybatis-dynamic-sql/blob/master/src/test/java/examples/simple/ReusableWhereTest.java

**Comment by a188037445** [comment]
Thank you, it's good!

**Comment by a188037445** [comment]
Sorry for reopening it, I just found a better way to do "count and paging".

There are some codes which need to be add first:
```java
@FunctionalInterface
public interface SelectInterface<T> {

    T select(BasicColumn... selectList);
}

public interface SqlBuilder {

    // only make it easier to build a lambda
    public static <T> SelectInterface<T> buildSelect(SelectInterface<T> si) {
        return si;
    }
}
```
Then we can do "count and paging" like this:
```java
    public void test() {
        SelectInterface<QueryExpressionDSL<SelectModel>.QueryExpressionWhereBuilder> si = buildSelect(c ->
                select(c)
                        .from(item)
                        .leftJoin(category)
                        .on(item.categoryId, equalTo(category.id))
                        .where(item.itemName, isLike("exampleName").then(s -> "%" + s + "%"))
                        .and(category.type, isEqualTo(123)));

        SelectStatementProvider ssp = si.select(count())
                .build()
                .render(RenderingStrategies.MYBATIS3);
        long total = itemMapper.count(ssp);

        ssp = si.select(item.id, item.itemName)
                .orderBy(item.id)
                .limit(10)
                .offset(0)
                .build()
                .render(RenderingStrategies.MYBATIS3);
        List<Map<String, Object>> map = commonSelectMapper.selectManyMappedRows(ssp);
    }
```
I think it's more easier than before, we don't need to write duplicate codes any more, do you think so?

**Comment by jeffgbutler** [comment]
I might challenge you to think a bit differently about it. You don't need to duplicate the `build` and `render` methods either!

Function composition is super powerful and you can do a lot with what's already in the library. For example, take a look at this as a solution.

This has a convenience interface called `WhereFinisher` but that's not totally necessary - it just makes things a bit easier to read (it's like a Kotlin type alias).

```java
@Test
void testCountAndPaging() {
    SelectStatementProvider ssp = buildSelect(count());

    assertThat(ssp.getSelectStatement()).isEqualTo(
            "select count(*) from AnimalData where id > #{parameters.p1,jdbcType=INTEGER}");
    assertThat(ssp.getParameters()).containsEntry("p1", 3);

    ssp = buildSelect(BasicColumn.columnList(id, bodyWeight, brainWeight),
        d -> d.orderBy(id).limit(10).offset(0));

    assertThat(ssp.getSelectStatement()).isEqualTo(
            "select id, body_weight, brain_weight " +
                    "from AnimalData " +
                    "where id > #{parameters.p1,jdbcType=INTEGER} " +
                    "order by id limit #{parameters.p2} offset #{parameters.p3}");
    assertThat(ssp.getParameters()).containsEntry("p1", 3);
    assertThat(ssp.getParameters()).containsEntry("p2", 10L);
    assertThat(ssp.getParameters()).containsEntry("p3", 0L);
}

private SelectStatementProvider buildSelect(BasicColumn... columns) {
    return buildSelect(columns, d -> d);
}

private SelectStatementProvider buildSelect(BasicColumn[] columns, WhereFinisher finisher) {
    return finisher.apply(
            select(columns)
                    .from(animalData)
                    .where(id, isGreaterThan(3)))
            .build()
            .render(RenderingStrategies.MYBATIS3);
}

@FunctionalInterface
interface WhereFinisher extends Function<QueryExpressionDSL<SelectModel>.QueryExpressionWhereBuilder,
        Buildable<SelectModel>> {}
```

**Comment by a188037445** [comment]
Yes, your code looks better.

Then my question is, how can I do this without any extra methods by myself? Because there may be a lot of similar logic, I don't like writing extra methods every time.

**Comment by jeffgbutler** [comment]
There are lots of ways you can deal with this. You can use the method you found, or my method, or you can think of others. It all depend on how abstract you want to it be.

My opinion is that making things more abstract makes them harder to understand. So I try to think about the future maintainers of my code and try to make it easier on them - that may mean a bit of duplication sometimes, but I think it's worth it.

The bottom line is that you can do it however seems best to you. But I don't think we need to add anything to the library for either of these methods. If you want to use your method, you can just put the interface and method in your own code and it will work perfectly fine.

**Comment by a188037445** [comment]
I respect your opinion, but I think it's necessary to improve this.

Maybe the code above is not elegant enough, now I have an improved version, only need to add a method:
```java
public static Function<QueryExpressionDSL.FromGatherer<SelectModel>, SelectStatementProvider> select(Function<QueryExpressionDSL.FromGatherer<SelectModel>, SelectStatementProvider> f) {
    return f;
}
```
Then we can use it like this:
```java
Function<QueryExpressionDSL.FromGatherer<SelectModel>, SelectStatementProvider> f = select(s ->
        s.from(item).build().render(RenderingStrategies.MYBATIS3));
SelectStatementProvider ssp = f.apply(select(count()));
SelectStatementProvider ssp = f.apply(select(...));
```
I think it is a good solution and not hard to unstandard, the reason for adding an empty method is to make it easier to build a lambda (otherwise we will need to write a long type name), it will be useful to many people, please think it over again.

**Comment by jeffgbutler** [comment]
You don't need the static function to make this work. Just make a very slight change...

```java
Function<QueryExpressionDSL.FromGatherer<SelectModel>, SelectStatementProvider> f =
    s -> s.from(item).build().render(RenderingStrategies.MYBATIS3);

SelectStatementProvider ssp = f.apply(select(count()));

SelectStatementProvider ssp = f.apply(select(...));
```

or this...

```java
Function<QueryExpressionDSL.FromGatherer<SelectModel>, QueryExpressionDSL<SelectModel>.QueryExpressionWhereBuilder> f =
    s -> s.from(item).where(item.itemName, isEqualTo("name"));

SelectStatementProvider ssp = f.apply(select(count())).build().render(RenderingStrategies.MYBATIS3);

SelectStatementProvider ssp = f.apply(select(item.id, item.itemName))
        .orderBy(item.id)
        .limit(10)
        .offset(0)
        .build()
        .render(RenderingStrategies.MYBATIS3);```

**Comment by mmm8955405** [comment]
> You don't need the static function to make this work. Just build the lambda directly...
> 
> ```java
> Function<QueryExpressionDSL.FromGatherer<SelectModel>, SelectStatementProvider> f =
>     s -> s.from(item).build().render(RenderingStrategies.MYBATIS3);
> 
> SelectStatementProvider ssp = f.apply(select(count()));
> 
> SelectStatementProvider ssp = f.apply(select(...));
> ```
> 
> or this...
> 
> ```java
> Function<QueryExpressionDSL.FromGatherer<SelectModel>, QueryExpressionDSL<SelectModel>.QueryExpressionWhereBuilder> f =
>     s -> s.from(item).where(item.itemName, isEqualTo("name"));
> 
> SelectStatementProvider ssp = f.apply(select(count())).build().render(RenderingStrategies.MYBATIS3);
> 
> SelectStatementProvider ssp = f.apply(select(item.id, item.itemName))
>         .orderBy(item.id)
>         .limit(10)
>         .offset(0)
>         .build()
>         .render(RenderingStrategies.MYBATIS3);
> ```

This expression is really too many. If a lot of link queries are added, the code readability is low

**Comment by mmm8955405** [comment]
![image](https://github.com/mybatis/mybatis-dynamic-sql/assets/5687907/e3f4f74b-c07e-410e-a324-981f721d72b3)

**Comment by mmm8955405** [comment]
Code generator, the default table name and entity name are the same, and the full class name must also be used
